pkg/apis/migration/v1alpha1/verify.go

```diff
@@ -0,0 +1,248 @@
+	UnhealthySourceNamespaces      = "UnhealthySourceNamespaces"
+	UnhealthyDestinationNamespaces = "UnhealthyDestinationNamespaces"
+)
+
+// Reasons
+const (
+	ErrorsDetected = "ErrorsDetected"
+)
+
+// Messages
+const (
+	UnhealthyDestination = "The destination cluster has unhealthy migrated namespaces []. See status.namespaces for details."
+	UnhealthySource      = "The source cluster has unhealthy namespaces []. See status.namespaces for details."
+)
+
+const (
+	// ImagePullBackOff is unhealthy container state
+	ImagePullBackOff = "ImagePullBackOff"
+	// CrashLoopBackOff is unhealthy container state
+	CrashLoopBackOff = "CrashLoopBackOff"
+)
+
+// UnhealthyResources is a store for unhealthy namespaces across clusters
+type UnhealthyResources struct {
+	Namespaces []UnhealthyNamespace `json:"namespaces,omitempty"`
+}
+
+// Workload is a store for unhealthy resource and it's dependents
+type Workload struct {
+	Name      string   `json:"name"`
+	Resources []string `json:"resources,omitempty"`
+}
+
+// UnhealthyNamespace is a store for unhealthy resources in a namespace
+type UnhealthyNamespace struct {
+	Name      string     `json:"name"`
+	Workloads []Workload `json:"workloads"`
+}
+
+// IsHealthy will report if there are any unhealthy namespaces were present
+func (m *MigMigration) IsHealthy() bool {
+	return m.Status.UnhealthyResources.Namespaces == nil || len(m.Status.UnhealthyResources.Namespaces) == 0
+}
+
+// VerifyPods would collect, display and notify about the pod health in a specific namespace
+func (u *UnhealthyResources) VerifyPods(client k8sclient.Client, options *k8sclient.ListOptions) (*[]unstructured.Unstructured, error) {
+	podList := corev1.PodList{}
+
+	err := client.List(context.TODO(), options, &podList)
+	if err != nil {
+		return nil, err
+	}
+
+	unhealthy := []unstructured.Unstructured{}
+	for _, pod := range podList.Items {
+		if pod.Status.Phase != corev1.PodFailed && !containerUnhealthy(pod) {
+			continue
+		}
+
+		obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&pod)
+		if err != nil {
+			return nil, err
+		}
+
+		unstructuredPod := unstructured.Unstructured{Object: obj}
+		unstructuredPod.SetKind("Pod")
+		unhealthy = append(unhealthy, unstructuredPod)
+	}
+
+	return &unhealthy, nil
+}
+
+func containerUnhealthy(pod corev1.Pod) bool {
+	// Consider suspicious containers as an error state
+	podContainersStatuses := append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...)
+	for _, containerStatus := range podContainersStatuses {
+		if containerStatus.Ready {
+			continue
+		}
+		terminationDetails := containerStatus.State.Terminated
+		if terminationDetails != nil && terminationDetails.ExitCode != 0 {
+			return true
+		}
+		waitingDetails := containerStatus.State.Waiting
+		if waitingDetails == nil {
+			continue
+		}
+		switch waitingDetails.Reason {
+		case ImagePullBackOff, CrashLoopBackOff:
+			break
+		default:
+			continue
+		}
+
+		return true
+	}
+
+	return false
+}
+
```

Comment:
I would expect these _find_ methods to be public and used by the controller, reconciler.

Reply:
Sure, could be fixed.